contracts/FeeCollector.sol

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+import { IFeeCollector } from "./interfaces/IFeeCollector.sol";
+import { LibTransfer } from "./lib/LibTransfer.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { Multicall } from "@openzeppelin/contracts/utils/Multicall.sol";
+
+/// @title A smart contract for registering vaults for payments.
+contract FeeCollector is IFeeCollector, Multicall, Ownable {
+    using LibTransfer for address payable;
+
+    address payable public guildTreasury;
+    uint96 public totalFeeBps;
+
+    mapping(string => FeeShare[]) internal feeSchemas;
+
+    Vault[] internal vaults;
+
+    /// @param guildTreasury_ The address that will receive Guild's share from the funds.
+    /// @param totalFeeBps_ The percentage of Guild's and any partner's share expressed in basis points.
+    constructor(address payable guildTreasury_, uint256 totalFeeBps_) {
+        guildTreasury = guildTreasury_;
+        totalFeeBps = uint96(totalFeeBps_);
+    }
+
+    function registerVault(address payable owner, address token, bool multiplePayments, uint128 fee) external {
+        Vault storage vault = vaults.push();
+        vault.owner = owner;
+        vault.token = token;
+        vault.multiplePayments = multiplePayments;
+        vault.fee = fee;
+
+        emit VaultRegistered(vaults.length - 1, owner, token, fee);
+    }
+
+    function payFee(uint256 vaultId) external payable {
+        if (vaultId >= vaults.length) revert VaultDoesNotExist(vaultId);
+
+        Vault storage vault = vaults[vaultId];
+
+        if (!vault.multiplePayments && vault.paid[msg.sender]) revert AlreadyPaid(vaultId, msg.sender);
+
+        uint256 requiredAmount = vault.fee;
+        vault.collected += uint128(requiredAmount);
+        vault.paid[msg.sender] = true;
+
+        // If the tokenAddress is zero, the payment should be in Ether, otherwise in ERC20.
+        address tokenAddress = vault.token;
+        if (tokenAddress == address(0)) {
+            if (msg.value != requiredAmount) revert IncorrectFee(vaultId, msg.value, requiredAmount);
+        } else {
+            if (msg.value != 0) revert IncorrectFee(vaultId, msg.value, 0);
+            payable(address(this)).sendTokenFrom(msg.sender, tokenAddress, requiredAmount);
```

Comment:
This requires a user to approve the tokens to be spent, so even if the choice is to forego checking allowance on the smart contract level, it could at the very least be helpful to mention this in the docs so that people making the UI can know to ensure that the user updates the respective ERC20's allowance before calling this.